src/ziplist.c

```diff
@@ -1498,10 +1498,96 @@ int ziplistValidateIntegrity(unsigned char *zl, size_t size, int deep,
     return 1;
 }
 
+/* Randomly select a pair of key and value.
+ * total_count is a pre-computed length/3 of the ziplist (to avoid calls to ziplistLen)
+ * 'key' and 'klen' are used to store the result key if it is a string. and
+ * 'klval' is used to store the key in case it is integer encoded.
+ * The other 3 arguments are used to store the value, and can be NULL if the
+ * value is not needed. */
+void ziplistRandomPair(unsigned char *zl,  unsigned long total_count,
+                       unsigned char **key, unsigned int *klen, long long *klval,
+                       unsigned char **value, unsigned int *vallen, long long *vlval)
+{
+    int ret;
+    unsigned char *p;
+
+    /* Generate even numbers, because ziplist saved K-V pair */
+    int r = (rand() % total_count) * 2;
+    p = ziplistIndex(zl, r);
+    ret = ziplistGet(p, key, klen, klval);
+    assert(ret != 0);
+
+    if (!value)
+        return;
+    p = ziplistNext(zl, p);
+    ret = ziplistGet(p, value, vallen, vlval);
+    assert(ret != 0);
+}
+
+/* int compare for qsort */
+int intCompare(const void *a, const void *b) {
+    return (*(int *) a - *(int *) b);
+}
+
+/* Helper method to store a string into from val or lval into dest */
+void ziplistSaveValue(unsigned char *val, unsigned int len, long long lval, sds *dest) {
+    if (val) {
+        *dest = sdsnewlen(val, len);
+        memcpy(*dest, val, len);
+        (*dest)[len] = '\0';
+    } else {
+        char buf[32];
+        int l = ll2string(buf,sizeof(buf),lval);
+        *dest = sdsnewlen(buf, l);
+    }
+}
+
+/* Randomly select unique count of key value pairs and store into 'keys' and
+ * 'vals' args. The order of the picked entries is random.
+ * The 'vals' arg can be NULL in which case we skip these. */
+void ziplistRandomPairs(unsigned char *zl, int count, sds *keys, sds *vals) {
```

Comment:
i think that by now, since we no longer follow the independent sds repo and others (rax, listpack, dict), we can allow ourselves to use sds in ziplist.
but your suggestion has the benefit that we can propagate integer values upwards, i'll look into that. thanks.